### PR TITLE
Generate Base Reconciler Code for Injection Based Controllers.

### PR DESCRIPTION
**Pull request by n3wscott**
There are a few cases where the setup of the controller and the wrapper around `Reconciler` are all the same. This aims to help reduce the amount of copy/paste code required to get a correct reconciler up and running. 

To enable generation of the Reconciler code, it is gated by a flag (**+genreconciler**) in the API structure:

```go

// +genclient
// +genreconciler
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object

// AddressableService is a Knative abstraction that encapsulates the interface by which Knative
// components express a desire to have a particular image cached.
type AddressableService struct {
...
```

This generates two things: a Reconciler/Controller pair that are intended to be imported and used, and the Stub of a ReconcilerImpl/ControllerImpl that import use the Reconciler/Controller with correct paths, intended to be copied out of the `./stub` directory.

See https://github.com/knative/sample-controller/pull/167 for full code changes.

Look at ApiServerSource for a more complex example: https://github.com/knative/eventing/pull/2487

### CODE REVIEW

File: codegen/cmd/reconciler-gen/generators/gen_stub_reconciler.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+// Check that our Reconciler implements Interface
+var _ Interface = (*Reconciler)(nil)
+
+// ReconcileKind implements Interface.ReconcileKind.
+func (r *Reconciler) ReconcileKind(ctx context.Context, o *{{.type|raw}}) error {
+	if o.GetDeletionTimestamp() != nil {
+		// Check for a DeletionTimestamp.  If present, elide the normal reconcile logic.
+		// When a controller needs finalizer handling, it would go here.
+		return nil
+	}
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_stub_reconciler.go]: Remove trailing whitespace:
```suggestion
	}
```

Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_stub_reconciler.go]: Remove trailing whitespace:
```suggestion
	}
```

File: codegen/cmd/reconciler-gen/args/args.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/args/args.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/gen_controller.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_controller.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/gen_reconciler.go [hunk 4]
```diff
@@ -0,0 +1,342 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_reconciler.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/gen_stub_controller.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_stub_controller.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/gen_stub_reconciler.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_stub_reconciler.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/generators.go [hunk 7]
```diff
@@ -0,0 +1,262 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/generators.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/main.go [hunk 8]
```diff
@@ -0,0 +1,39 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/main.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/gen_stub_reconciler.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+// Check that our Reconciler implements Interface
+var _ Interface = (*Reconciler)(nil)
+
+// ReconcileKind implements Interface.ReconcileKind.
+func (r *Reconciler) ReconcileKind(ctx context.Context, o *{{.type|raw}}) {{.reconcilerEvent|raw}} {
+	if o.GetDeletionTimestamp() != nil {
+		// Check for a DeletionTimestamp.  If present, elide the normal reconcile logic.
+		// When a controller needs finalizer handling, it would go here.
+		return nil
+	}
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_stub_reconciler.go]: Remove trailing whitespace:
```suggestion
	}
```

File: codegen/cmd/reconciler-gen/args/args.go (hunk 2 above)
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/args/args.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/gen_controller.go (hunk 3 above)
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_controller.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/gen_reconciler.go [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_reconciler.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/gen_stub_controller.go (hunk 5 above)
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_stub_controller.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/gen_stub_reconciler.go [hunk 11]
```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_stub_reconciler.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/generators.go (hunk 7 above)
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/generators.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/main.go (hunk 8 above)
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/main.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/gen_stub_reconciler.go (hunk 9 above)
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_stub_reconciler.go]: Remove trailing whitespace:
```suggestion
	}
```

File: codegen/cmd/reconciler-gen/args/args.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/args/args.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/gen_controller.go [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_controller.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/gen_reconciler.go (hunk 10 above)
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_reconciler.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/gen_stub_controller.go (hunk 5 above)
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_stub_controller.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/gen_stub_reconciler.go (hunk 11 above)
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/gen_stub_reconciler.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/generators.go [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/generators.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/reconciler-gen/generators/namesystems.go [hunk 15]
```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2019 The Knative Authors.
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/namesystems.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/reconciler-gen/generators/packages.go [hunk 16]
```diff
@@ -0,0 +1,429 @@
+/*
+Copyright 2019 The Knative Authors.
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/generators/packages.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/reconciler-gen/main.go [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
```
Comment (mattmoor) [inline codegen/cmd/reconciler-gen/main.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: codegen/cmd/injection-gen/generators/reconciler_reconciler_stub.go [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+// Check that our Reconciler implements Interface
+var _ Interface = (*Reconciler)(nil)
+
+// ReconcileKind implements Interface.ReconcileKind.
+func (r *Reconciler) ReconcileKind(ctx context.Context, o *{{.type|raw}}) {{.reconcilerEvent|raw}} {
+	if o.GetDeletionTimestamp() != nil {
+		// Check for a DeletionTimestamp.  If present, elide the normal reconcile logic.
+		// When a controller needs finalizer handling, it would go here.
+		return nil
+	}
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/reconciler_reconciler_stub.go]: Remove trailing whitespace:
```suggestion
	}
```

File: codegen/cmd/injection-gen/generators/reconciler_reconciler_stub.go [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/reconciler_reconciler_stub.go]: Remove trailing whitespace:
```suggestion
	}
```

Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/reconciler_reconciler_stub.go]: Remove trailing whitespace:
```suggestion
	}
```

File: codegen/cmd/injection-gen/generators/reconciler_reconciler_stub.go [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+	r.Core = *core
+}
+
+// ReconcileKind implements Interface.ReconcileKind.
+func (r *Reconciler) ReconcileKind(ctx context.Context, o *{{.type|raw}}) {{.reconcilerEvent|raw}} {
+	if o.GetDeletionTimestamp() != nil {
+		// Check for a DeletionTimestamp.  If present, elide the normal reconcile logic.
+		// When a controller needs finalizer handling, it would go here.
+		return nil
+	}
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/reconciler_reconciler_stub.go]: Remove trailing whitespace:
```suggestion
	}
```

File: codegen/cmd/injection-gen/generators/reconciler_reconciler_stub.go [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/reconciler_reconciler_stub.go]: Remove trailing whitespace:
```suggestion
	}
```

File: codegen/cmd/injection-gen/generators/reconciler_reconciler.go [hunk 22]
```diff
@@ -0,0 +1,352 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/reconciler_reconciler.go]: Knative

File: codegen/cmd/injection-gen/generators/reconciler_reconciler.go [hunk 23]
```diff
@@ -0,0 +1,352 @@
...
+	logger := {{.loggingFromContext|raw}}(ctx)
+
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := {{.cacheSplitMetaNamespaceKey|raw}}(key)
+	if err != nil {
+		logger.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+
+    // TODO: this is needed for serving.
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/reconciler_reconciler.go]: Seems important?

Comment (n3wscott) [reply codegen/cmd/injection-gen/generators/reconciler_reconciler.go]: Serving has a bunch of other problems, there will be updates to this code as more integration happens.

File: codegen/cmd/injection-gen/generators/client.go [hunk 24]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2019 The Kubernetes Authors.
+Copyright 2019 The Knative Authors.
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/client.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakeclient.go [hunk 25]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2019 The Kubernetes Authors.
+Copyright 2019 The Knative Authors.
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakeclient.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakeduck.go [hunk 26]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2019 The Kubernetes Authors.
+Copyright 2019 The Knative Authors.
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakeduck.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakefactory.go [hunk 27]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2019 The Kubernetes Authors.
+Copyright 2019 The Knative Authors.
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakefactory.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakeinformer.go [hunk 28]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2019 The Kubernetes Authors.
+Copyright 2019 The Knative Authors.
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakeinformer.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/reconciler_reconciler.go [hunk 29]
```diff
@@ -0,0 +1,352 @@
...
+	resource := original.DeepCopy()
+
+	// Reconcile this copy of the resource and then write back any status
+	// updates regardless of whether the reconciliation errored out.
+	reconcileEvent := r.reconciler.ReconcileKind(ctx, resource)
+
+	// Synchronize the finalizers.
+	if equality.Semantic.DeepEqual(original.Finalizers, resource.Finalizers) {
+		// If we didn't change finalizers then don't call updateFinalizers.
+	} else if _, updated, fErr := r.updateFinalizers(ctx, resource); fErr != nil {
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/reconciler_reconciler.go]: This finalizer handling is racy.  I'd rather we rip out all of the finalizer bits here until we introduce an interface that lets us detect the intent to handle finalizers properly.

File: codegen/cmd/injection-gen/generators/client.go (hunk 24 above)
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/client.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakeclient.go (hunk 25 above)
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakeclient.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakeduck.go (hunk 26 above)
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakeduck.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakefactory.go (hunk 27 above)
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakefactory.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakeinformer.go (hunk 28 above)
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakeinformer.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/reconciler_reconciler.go (hunk 29 above)
Comment (n3wscott) [reply codegen/cmd/injection-gen/generators/reconciler_reconciler.go]: it is now commented out and we will do a followup to add it back in. There are some ideas around reflecting the given reconciler to help understand what to do

File: codegen/cmd/injection-gen/generators/client.go (hunk 24 above)
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/client.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakeclient.go (hunk 25 above)
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakeclient.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakeduck.go (hunk 26 above)
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakeduck.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakefactory.go (hunk 27 above)
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakefactory.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakeinformer.go (hunk 28 above)
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakeinformer.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/client.go (hunk 24 above)
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/client.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakeclient.go (hunk 25 above)
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakeclient.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakeduck.go (hunk 26 above)
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakeduck.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakefactory.go (hunk 27 above)
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakefactory.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/fakeinformer.go (hunk 28 above)
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/fakeinformer.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors.
```

File: codegen/cmd/injection-gen/generators/reconciler_reconciler.go (hunk 29 above)
Comment (n3wscott) [reply codegen/cmd/injection-gen/generators/reconciler_reconciler.go]: Removed.

**Pull request merged**